Numerous small updates

bedder opened the pull request.

Commit:
Misc updates

Commit:
Fixing complex dtype definition

bedder commented on numpy-stubs/__init__.pyi:
```diff
@@ -77,6 +77,7 @@ _DtypeLike = Union[
     Dict[str, Tuple[_DtypeLikeNested, int]],
     # (base_dtype, new_dtype)
     Tuple[_DtypeLikeNested, _DtypeLikeNested],
+    Sequence[Tuple[str, _DtypeLikeNested]],
```
Support for things like:
```
np.array(datetime, [('foo', np.float32), ('bar', 'U')]
```

bedder commented on numpy-stubs/__init__.pyi:
```diff
@@ -389,7 +390,7 @@ class _real_generic(generic):
 
 class number(generic):
-    def __init__(self, value: Union[SupportsInt, SupportsFloat] = ...) -> None: ...
+    def __init__(self, value: Union[SupportsInt, SupportsFloat, str] = ...) -> None: ...
```
Support for things like:
```
np.float32("1.23")
```

bedder reviewed (commented):
Notes for reviewers.

elvijs commented on numpy-stubs/__init__.pyi:
```diff
@@ -450,6 +451,11 @@ def array(
     subok: bool = ...,
     ndmin: int = ...,
 ) -> ndarray: ...
+def asarray(
```
The lack of a newline is really triggering my OCD here..

bedder replied:
I'm sorry to say that the official numpy stubs repo uses this formatting, plus [black](https://github.com/python/black) for enforcing consistent formatting.

We could do nicer formatting/change the rules for how we want to format things... but then it would make it more difficult if we wanted to try to push things upstream to the official repo.

elvijs replied:
Oh no, this is filthy! But agreed, let's stick with the project conventions

elvijs commented on numpy-stubs/__init__.pyi:
```diff
@@ -450,6 +451,11 @@ def array(
 ) -> ndarray: ...
+def asarray(
+    a: Union[SupportsFloat, Sequence, ndarray],
+    dtype: _DtypeLike = ...,
```
Should this be optional?

bedder replied:
Ah, yes. Good catch!

(I think I got a bit confused w.r.t. the stub format for saying that something has a default value. Will make the change.)

elvijs commented on numpy-stubs/__init__.pyi:
```diff
@@ -450,6 +451,11 @@ def array(
+def asarray(
+    a: Union[SupportsFloat, Sequence, ndarray],
+    dtype: _DtypeLike = ...,
+    order: str = ...,
```
Should this be optional? Looking at https://docs.scipy.org/doc/numpy/reference/generated/numpy.asarray.html

elvijs commented on numpy-stubs/__init__.pyi:
```diff
@@ -450,6 +451,11 @@ def array(
     subok: bool = ...,
     ndmin: int = ...,
 ) -> ndarray: ...
+def asarray(
```
Is it possible to add some tests for this?

bedder replied:
Done.

elvijs commented on numpy-stubs/core/records/__init__.pyi:
```diff
@@ -0,0 +1,14 @@
+
+import numpy as np
+
+def fromarrays(
```
Can we add some tests for this?

bedder replied:
Done.

elvijs reviewed (commented):
A few minor comments. Would be good to address before merging

Commit:
Refining type stubs, adding tests

elvijs reviewed (approved):
Looks good! The only minor suggestion I have is a few "fail" tests, but equally happy for you to merge as-is :)

bedder commented:
> Looks good! The only minor suggestion I have is a few "fail" tests, but equally happy for you to merge as-is :)

Failure cases are a bit unusual in the testing style used here, so I'm not sure how to do useful ones - I found it useful for cases where overload are used (e.g. checking that with the [datetime64 constructor overload](https://github.com/Prowler-io/numpy-stubs/blob/master/numpy-stubs/__init__.pyi#L399) can can't call it with [plausible-looking arguments that match no overloads](https://github.com/Prowler-io/numpy-stubs/blob/master/tests/fail/scalars.py#L11)), but can't see what value it would bring for these new hints (other than checking that arguments of the wrong type do cause errors to be raised).

I'd be happy to hear about any failure cases you think might be useful for this!

elvijs committed:
Little workaround

Commit:
Added toxfile to assist with testing

Commit:
Fixed test failure (where mypy appears to be getting confused?)

elvijs committed:
Fix dtype type hint

List is an invariant type, so the substitution rules
were not achieving the desired result

elvijs committed:
Added some failure tests
